### PR TITLE
Fix search params being dropped when redirected to non-deck path

### PR DESCRIPTION
**Pull request by renchap**
For example `http://localhost:3000/deck/directory?order=active&local=false` to `http://localhost:3000/directory?order=active&local=false`

This works when you are logged-in but do not have the advanced interface enabled.

When you are logged out the redirect is done by Rails (in `PermalinkRedirector`) and the query string is not kept, I am not sure how to fix it.

### CODE REVIEW

**Comment by ClearlyClaire** [comment]
Maybe calling `PermalinkRedirector` with `request.original_fullpath` instead of `request.path`?

**Comment by renchap** [comment]
Updated with this change, and added some specs. It is a bit weird that the other cases return a full uri for `redirect_path` but here we have a path, for the redirector calls `url_for` if the URL is for a AP object.

I also changed `path_segments` to only use what precedes a `?` so `second_segment` never contains a query string, like it previously did. This is not mandatory and it works even if the query string is kept (not sure why, `find_by(id: …)` strips it?) but it feels cleaner this way.

File: app/javascript/mastodon/features/ui/index.jsx [hunk 1]
```diff
@@ -186,7 +186,7 @@ class SwitchingColumnsArea extends PureComponent {
             {redirect}
 
             {singleColumn ? <Redirect from='/deck' to='/home' exact /> : null}
-            {singleColumn && pathName.startsWith('/deck/') ? <Redirect from={pathName} to={pathName.slice(5)} /> : null}
+            {singleColumn && pathName.startsWith('/deck/') ? <Redirect from={pathName} to={{...this.props.location, pathname: pathName.slice(5)}} /> : null}
```
Comment (ClearlyClaire) [inline app/javascript/mastodon/features/ui/index.jsx:189]: Hm, do we really want to use `this.props.location` in here? iirc it is mutable, would that always behave as expected?

Comment (renchap) [reply app/javascript/mastodon/features/ui/index.jsx:189]: I am not sure what else we can use here. The props is injected by `react-router`, so it should contain the latest value?

Comment (ClearlyClaire) [reply app/javascript/mastodon/features/ui/index.jsx:189]: Hm, I guess this is fine. The `pathName` already comes from there anyway. Though I do find the whole pattern a little weird.

**Review by ClearlyClaire — APPROVED** [review]

**Pull request merged**